### PR TITLE
thrift: Disable enveloping for TChannel only

### PR DESCRIPTION
**Pull request by abhinav**
(This supersedes #256)

This disables enveloping of Thrift requests and responses for TChannel while
leaving it enabled for everything else. It uses the transport options
functionality to propagate this requirement from the TChannel transport to the
Thrift encoding.

### CODE REVIEW

**Comment by abhinav** [comment]
@yarpc/golang This is ready for review. New crossdock tests incoming.

**Comment by breerly** [comment]
screen shot 2016-07-14 at 4 10 57 pm

File: crossdock/client/echo/thrift.go [hunk 1]
```diff
@@ -42,7 +43,7 @@ func Thrift(t crossdock.T) {
 	fatals.NoError(dispatcher.Start(), "could not start Dispatcher")
 	defer dispatcher.Stop()
 
-	client := echoclient.New(dispatcher.Channel("yarpc-test"))
+	client := echoclient.New(dispatcher.Channel("yarpc-test"), thrift.DisableEnveloping)
```
Comment (breerly) [inline crossdock/client/echo/thrift.go]: Please add a comment detailing why we need to disable enveloping here. Also, a TODO to remove when we have enveloping support across langs is probably in order.

Comment (breerly) [inline crossdock/client/echo/thrift.go]: I would probably add the same TODO to each of these Crossdock tests that require enveloping disabled.

File: encoding/thrift/inbound.go [hunk 2]
```diff
@@ -48,7 +49,16 @@ func (t thriftHandler) Handle(ctx context.Context, _ transport.Options, treq *tr
 		return err
 	}
 
-	envelope, err := t.Protocol.DecodeEnveloped(bytes.NewReader(body))
+	// We disable enveloping if either the client or the transport requires it.
+	proto := t.Protocol
+	if t.DisableEnveloping || isEnvelopingDisabled(opts) {
```
Comment (breerly) [inline encoding/thrift/inbound.go]: `isEnvelopingDisabled` is super clean - 👍

File: transport/tchannel/handler.go [hunk 3]
```diff
@@ -142,7 +140,7 @@ func (h handler) callHandler(ctx context.Context, call inboundCall) error {
 		return err
 	}
 
-	return h.Handler.Handle(ctx, tchannelOptions, treq, rw)
+	return h.Handler.Handle(ctx, transportOptions, treq, rw)
```
Comment (breerly) [inline transport/tchannel/handler.go]: So `transportOptions` is where we specify the default options, including requiring enveloping disabled.

File: encoding/thrift/transport_options.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+// Use this from a transport implementation that wishes to disable Thrift
+// enveloping.
+//
+// 	func (myOutbound) Options() transport.Options {
+// 		return otherOptions.Merge(thrift.DisableEnvelopingForTransport)
+// 	}
+var DisableEnvelopingForTransport transport.Options
+
+func init() {
+	DisableEnvelopingForTransport = transport.Options{}.With(
```
Comment (breerly) [inline encoding/thrift/transport_options.go]: Why run this in `init` instead of when declaring `DisableEnvelopingForTransport`?

**Comment by breerly** [comment]
LGTM (beside the few Q's/nits and questionable branch name)

File: encoding/thrift/transport_options.go (hunk 4 above)
Comment (prashantv) [inline encoding/thrift/transport_options.go]: This is to make the `godoc` output nicer. It'll show the exported global variables with their declarations, so putting it into init avoids the noise in the documentation

File: crossdock/client/echo/thrift.go (hunk 1 above)
Comment (prashantv) [inline crossdock/client/echo/thrift.go]: +1 for comments

**Comment by prashantv** [comment]
lgtm

File: encoding/thrift/transport_options.go (hunk 4 above)
Comment (breerly) [inline encoding/thrift/transport_options.go]: ty sir - good to know :)

File: transport/tchannel/handler.go (hunk 3 above)
Comment (abhinav) [inline transport/tchannel/handler.go]: Sort-of. Transport options is where transports specify additional
requirements. Right now, TChannel is the only transport that has non-standard
requirements, and they're constant: "Don't envelope Thrift payloads."

If in the future we add more options for TChannel, `transportOptions` may
become,

    transport.Options{}.Merge(
        thrift.DisableEnvelopingForTransport,
        other.OptionA(..),
        OptionB(..),
    )

File: crossdock/client/echo/thrift.go (hunk 1 above)
Comment (abhinav) [inline crossdock/client/echo/thrift.go]: Added comments

**Comment by breerly** [comment]
lgtm

**Pull request merged**